### PR TITLE
Use pytest to run the smoke test against the live api

### PR DESCRIPTION
**Pull request by miaucl**
Use the same framework as the unit tests to run the smoke test against the live api. The tests are split into login and methods in preparation to be able to run this smoke test also on forked PRs in a safe manner. More about it later...

Further changes required to prepare for this are minor adjustments to restore the auth information to the bring instance (not a "public" feature for now, but might become one, as currently all the required information to restore an "auth state" is contained in the headers (after adding the `public_uuid` as well^^)). Further, adding helpers to persist the auth state for short durations (ignoring the refresh token and expiration date).

### CODE REVIEW

**Comment by tr4nt0r** [comment]
vscode test discovery seems to have problems finding the tests

File: smoke_test/test_1_setup_headers.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+from bring_api.exceptions import BringEMailInvalidException, BringUserUnknownException
+from smoke_test.conftest import save_headers
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TestSetupAndLogin:
+    """Test setup and login."""
+
+    async def test_does_user_exist(self, bring_no_auth: Bring):
```
Comment (tr4nt0r) [inline smoke_test/test_1_setup_headers.py:18]:
```suggestion
    async def test_does_user_exist(self, bring_no_auth: Bring) -> None:
```

File: smoke_test/test_2_methods.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+load_dotenv()
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class TestMethods:
+    """Test methods."""
+
+    async def test_add_complete_remove(self, bring: Bring, test_list: BringList):
```
Comment (tr4nt0r) [inline smoke_test/test_2_methods.py:19]:
```suggestion
    async def test_add_complete_remove(self, bring: Bring, test_list: BringList) -> None:
```

File: smoke_test/test_2_methods.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+        _LOGGER.info("List recently items: %s", items["recently"])
+
+        # Remove an item from a list
+        await bring.remove_item(test_list["listUuid"], "Äpfel")
+
+        # Get all the items of a list
+        items = await bring.get_list(test_list["listUuid"])
+        _LOGGER.info("List all items: %s / %s", items["purchase"], items["recently"])
+
+    async def test_push_notifications(self, bring: Bring, test_list: BringList):
```
Comment (tr4nt0r) [inline smoke_test/test_2_methods.py:43]:
```suggestion
    async def test_push_notifications(self, bring: Bring, test_list: BringList) -> None:
```

File: smoke_test/test_2_methods.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+        # Send a going shopping notification
+        await bring.notify(test_list["listUuid"], BringNotificationType.GOING_SHOPPING)
+
+        # Send a urgent message with argument item name
+        await bring.notify(
+            test_list["listUuid"], BringNotificationType.URGENT_MESSAGE, "Pouletbrüstli"
+        )
+
+    async def test_translation(self, bring: Bring, test_list: BringList):
```
Comment (tr4nt0r) [inline smoke_test/test_2_methods.py:54]:
```suggestion
    async def test_translation(self, bring: Bring, test_list: BringList) -> None:
```

File: smoke_test/test_2_methods.py [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+            _LOGGER.info("Item: %s, translation: %s", v, item)
+
+            await bring.remove_item(test_list["listUuid"], k)
+
+        # reset locale to original value for other tests
+        bring.user_list_settings[test_list["listUuid"]][
+            "listArticleLanguage"
+        ] = locale_org
+
+    async def test_batch_list_operations(self, bring: Bring, test_list: BringList):
```
Comment (tr4nt0r) [inline smoke_test/test_2_methods.py:98]:
```suggestion
    async def test_batch_list_operations(self, bring: Bring, test_list: BringList) -> None:
```

File: smoke_test/test_2_methods.py [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                "operation": "REMOVE",
+            },
+        ]
+        await bring.batch_update_list(test_list["listUuid"], add_complete_delete_items)
+        _LOGGER.info(
+            "Submit batch update with ADD, COMPLETE, REMOVE operations to list %s",
+            test_list["name"],
+        )
+
+    async def test_article_language(self, bring: Bring, test_list: BringList):
```
Comment (tr4nt0r) [inline smoke_test/test_2_methods.py:193]:
```suggestion
    async def test_article_language(self, bring: Bring, test_list: BringList) None:
```

File: smoke_test/conftest.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+load_dotenv()
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def save_headers(headers: dict[str, str]) -> None:
+    """Save the headers locally."""
+    with open(".headers", "w") as file:
+        file.write(headers_serialize(headers))
```
Comment (tr4nt0r) [inline smoke_test/conftest.py:22]:
```suggestion
    with open(".headers", "w", encoding="utf-8") as file:
        file.write(headers_serialize(headers))
```

File: smoke_test/conftest.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+
+def load_headers() -> dict[str, str]:
+    """Load the headers locally."""
+    # Open and read the file
+    with open(".headers") as file:
+        return headers_deserialize(file.read() or "{}")
```
Comment (tr4nt0r) [inline smoke_test/conftest.py:29]:
```suggestion
    with open(".headers", encoding="utf-8") as file:
        return headers_deserialize(file.read() or "{}")
```

**Review by tr4nt0r — CHANGES_REQUESTED** [review]

**Comment by miaucl** [comment]
> vscode test discovery seems to have problems finding the tests
> 
> 
> 
> Adding the folder in pytest settings might help:
> 
> 
> 
> ```
> 
> testpaths = [
> 
>     "tests",
> 
>     "smoke_test"
> 
> ]
> 
> ```
> 
> 
> 
> also, why so much log output in the cli?

I dont want to run it by default, which is why it is not in there. Only the smoke test should run it!

**Comment by miaucl** [comment]
> vscode test discovery seems to have problems finding the tests
> 
> 
> 
> Adding the folder in pytest settings might help:
> 
> 
> 
> ```
> 
> testpaths = [
> 
>     "tests",
> 
>     "smoke_test"
> 
> ]
> 
> ```
> 
> 
> 
> also, why so much log output in the cli?

Verbose output is good to have. Usually one does not care about unit test. Except when they fail and them you want as much info as possible, no?

**Comment by tr4nt0r** [comment]
> > vscode test discovery seems to have problems finding the tests
> > Adding the folder in pytest settings might help:
> > ```
> > 
> > testpaths = [
> > 
> >     "tests",
> > 
> >     "smoke_test"
> > 
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > also, why so much log output in the cli?
> 
> Verbose output is good to have. Usually one does not care about unit test. Except when they fail and them you want as much info as possible, no?

But then i would maybe run it with very verbose enabled or in debug (in which having tests autodiscovered in vscode would help)

**Comment by miaucl** [comment]
Smoke test temporarily enabled also on this branch runs with success

[Smoke Test](https://github.com/miaucl/bring-api/actions/workflows/smoke-test.yaml)

_Disabled again for this branch, it will only run on master with this workflow. More to come..._

**Review by tr4nt0r — APPROVED** [review]

**Pull request merged**